### PR TITLE
feat: track persistence metrics

### LINKED ISSUE #10220: track persistence duration
ref #9794

we want metrics how long persistence took and should track this two fold:

1. with a histogram here https://github.com/paradigmxyz/reth/blob/c4aae6db361d2ac33fed18f25bb50158907723ee/crates/engine/tree/src/tree/metrics.rs#L9-L9 how long this took: https://github.com/paradigmxyz/reth/blob/c4aae6db361d2ac33fed18f25bb50158907723ee/crates/engine/tree/src/tree/mod.rs#L1980-L1980 this can be done by also storing an `Instant` together with the receiver.
2. and actual metrics for how long these took https://github.com/paradigmxyz/reth/blob/c4aae6db361d2ac33fed18f25bb50158907723ee/crates/engine/tree/src/persistence.rs#L63-L63 which should be stored in there own metrics struct

### LINKED ISSUE #10221: extract persistence event handling into standalone functions
for readability it would be nice to have this code in standalone functions

https://github.com/paradigmxyz/reth/blob/c4aae6db361d2ac33fed18f25bb50158907723ee/crates/engine/tree/src/persistence.rs#L63-L73

like `on_remove_blocks`, `on_save_blocks` etc.

### PR DESCRIPTION
**Pull request by malik672**
closes #10220
closes https://github.com/paradigmxyz/reth/issues/10221

### CODE REVIEW

File: crates/engine/tree/src/persistence.rs [hunk 1]
```diff
@@ -58,44 +67,67 @@ where
...
                 }
             }
         }
         Ok(())
     }
+
+    fn on_remove_blocks_above(
+        &self,
+        new_tip_num: u64,
+        sender: oneshot::Sender<()>,
```
Comment (mattsse) [inline crates/engine/tree/src/persistence.rs:93]: this doesn't need the sender as argument

File: crates/engine/tree/src/persistence.rs [hunk 2]
```diff
@@ -58,44 +67,67 @@ where
...
+        // we ignore the error because the caller may or may not care about the result
+        let _ = sender.send(());
+        self.metrics.remove_blocks_above_duration.record(start_time.elapsed());
+        Ok(())
+    }
+
+    fn on_save_blocks(
+        &self,
+        blocks: Vec<ExecutedBlock>,
+        sender: oneshot::Sender<Option<B256>>,
```
Comment (mattsse) [inline crates/engine/tree/src/persistence.rs:111]: same here, the function should return the Option and then caller should send it

**Review by mattsse — CHANGES_REQUESTED** [review]
some nits,

**Comment by mattsse** [comment]
@mvares this also closes #10221 as a drive by

**Comment by mvares** [comment]
> @mvares this also closes #10221 as a drive by

All fine! Yesterday I had some problems in my computer, but I already fixed it. Thanks!

File: crates/engine/tree/src/persistence.rs [hunk 3]
```diff
@@ -58,44 +67,54 @@ where
...
-                    let last_block_hash = last_block.block().hash();
-
-                    let provider_rw = self.provider.provider_rw()?;
-                    let static_file_provider = self.provider.static_file_provider();
-
-                    UnifiedStorageWriter::from(&provider_rw, &static_file_provider)
-                        .save_blocks(&blocks)?;
-                    UnifiedStorageWriter::commit(provider_rw, static_file_provider)?;
-
+                    let result = self.on_save_blocks(blocks);
```
Comment (mattsse) [inline crates/engine/tree/src/persistence.rs:75]:
```suggestion
                    let result = self.on_save_blocks(blocks)?;
```

File: crates/engine/tree/src/persistence.rs [hunk 4]
```diff
@@ -58,44 +67,54 @@ where
...
                     // we ignore the error because the caller may or may not care about the result
-                    let _ = sender.send(Some(last_block_hash));
+                    let _ = sender.send(result.ok().flatten());
                 }
                 PersistenceAction::PruneBefore(block_num, sender) => {
+                    let start_time = Instant::now();
```
Comment (mattsse) [inline crates/engine/tree/src/persistence.rs:80]: let's also move this into prune_before, so that we're consistent with where we track metrics

**Review by mattsse — CHANGES_REQUESTED** [review]
last nits

File: crates/engine/tree/src/persistence.rs (hunk 4 above)
Comment (malik672) [reply crates/engine/tree/src/persistence.rs]: @mattsse done

Comment (malik672) [reply crates/engine/tree/src/persistence.rs]: @mattsse what's latency metrics for?

File: crates/engine/tree/src/persistence.rs [hunk 5]
```diff
@@ -96,6 +89,33 @@ where
...
+
+        self.metrics.remove_blocks_above_duration.record(start_time.elapsed());
+        Ok(())
+    }
+
+    fn on_save_blocks(&self, blocks: Vec<ExecutedBlock>) -> Result<Option<B256>, PersistenceError> {
+        let start_time = Instant::now();
+        let last_block_hash = blocks.last().map(|block| block.block().hash());
+
+        if let Some(last_block_hash) = last_block_hash {
```
Comment (mattsse) [inline crates/engine/tree/src/persistence.rs]:
```suggestion
        if last_block_hash.is_some() {
```

File: crates/engine/tree/src/tree/mod.rs [hunk 6]
```diff
@@ -2004,15 +2001,22 @@ impl PersistenceState {
...
     /// Sets state for a finished persistence task.
-    fn finish(&mut self, last_persisted_block_hash: B256, last_persisted_block_number: u64) {
+    fn finish(
+        &mut self,
+        last_persisted_block_hash: B256,
+        last_persisted_block_number: u64,
+    ) -> Duration {
         trace!(target: "engine", block= %last_persisted_block_number, hash=%last_persisted_block_hash, "updating persistence state");
+        let duration =
+            self.rx.take().map(|(_, start_time)| start_time.elapsed()).unwrap_or_default();
```
Comment (mattsse) [inline crates/engine/tree/src/tree/mod.rs:2015]: I think this is fine

**Review by mattsse — APPROVED** [review]
lgtm, pending @Rjected @fgimenez

File: crates/engine/tree/src/tree/mod.rs [hunk 7]
```diff
@@ -2004,15 +2001,22 @@ impl PersistenceState {
...
         self.rx = None;
```
Comment (Rjected) [inline crates/engine/tree/src/tree/mod.rs:2016]: not necessary if we `take` rx right?

**Review by Rjected — APPROVED** [review]
lgtm, just one comment on removing `self.rx = None` after `take`

File: crates/engine/tree/src/tree/mod.rs (hunk 7 above)
Comment (malik672) [reply crates/engine/tree/src/tree/mod.rs]: removed

File: crates/engine/tree/src/tree/mod.rs (hunk 6 above)
Comment (malik672) [reply crates/engine/tree/src/tree/mod.rs:2015]: done

**Review by shekhirin — CHANGES_REQUESTED** [review]

File: crates/engine/tree/src/tree/metrics.rs [hunk 8]
```diff
@@ -15,5 +15,7 @@ pub(crate) struct EngineApiMetrics {
     pub(crate) forkchoice_updated_messages: Counter,
     /// The total count of new payload messages received.
     pub(crate) new_payload_messages: Counter,
+    /// Histogram of persistence operation durations (in seconds)
+    pub(crate) persistence_duration: Histogram,
```
[the remark this answers is not in the document]
Comment (mattsse) [reply crates/engine/tree/src/tree/metrics.rs:19]: measuring this in seconds is not useful

File: crates/engine/tree/src/tree/mod.rs [hunk 9]
```diff
@@ -2004,15 +2001,21 @@ impl PersistenceState {
...
+        last_persisted_block_hash: B256,
+        last_persisted_block_number: u64,
+    ) -> Duration {
         trace!(target: "engine", block= %last_persisted_block_number, hash=%last_persisted_block_hash, "updating persistence state");
-        self.rx = None;
+        let duration =
+            self.rx.take().map(|(_, start_time)| start_time.elapsed()).unwrap_or_default();
         self.last_persisted_block_number = last_persisted_block_number;
         self.last_persisted_block_hash = last_persisted_block_hash;
+        duration
```
[the remark this answers is not in the document]
Comment (mattsse) [reply crates/engine/tree/src/tree/mod.rs]: you're correct we've already used this

File: crates/engine/tree/src/tree/mod.rs [hunk 10]
```diff
@@ -853,32 +853,29 @@ where
...
-                        return Ok(())
-                    };
-                    if let Some(block) =
-                        self.state.tree_state.block_by_hash(last_persisted_block_hash)
-                    {
-                        self.persistence_state.finish(last_persisted_block_hash, block.number);
-                        self.on_new_persisted_block();
-                    } else {
-                        error!("could not find persisted block with hash {last_persisted_block_hash} in memory");
+            if let Some((mut rx, start_time)) = self.persistence_state.rx.take() {
```
[the remark this answers is not in the document]
Comment (mattsse) [reply crates/engine/tree/src/tree/mod.rs]: right, added the expect back

**Review by mattsse — APPROVED** [review]
pending @shekhirin

**Review by shekhirin — APPROVED** [review]

**Pull request merged**